monai/transforms/utils.py

```diff
@@ -1660,5 +1661,65 @@ def sync_meta_info(key, data_dict, t: bool = True):
     return d
 
 
+def check_boundaries(boundaries) -> None:
+    """
+    Check boundaries for Signal transforms
+    """
+    if not (
+        isinstance(boundaries, Sequence) and len(boundaries) == 2 and all(isinstance(i, float) for i in boundaries)
+    ):
+        raise ValueError("Incompatible values: boundaries needs to be a list of float.")
+
+
+def paste_slices(tup):
+    """
+    given a tuple (pos,w,max_w), return a tuple of slices
+    """
+    pos, w, max_w = tup
+    max_w = max_w.shape[len(max_w.shape) - 1]
+    orig_min = max(pos, 0)
+    orig_max = min(pos + w, max_w)
+    block_min = -min(pos, 0)
+    block_max = max_w - max(pos + w, max_w)
+    block_max = block_max if block_max != 0 else None
+    return slice(orig_min, orig_max), slice(block_min, block_max)
+
+
+def paste(orig, block, loc):
+    """
+    given a location (loc) and an original array (orig), paste a block array into it
+    """
+    loc_zip = zip(loc, block.shape, orig)
+    orig_slices, block_slices = zip(*map(paste_slices, loc_zip))
+
+    orig[:, orig_slices[0]] = block[block_slices[0]]
+
+    if orig.shape[0] == 1:
+        orig = orig.squeeze()
+    return orig
+
+
+def squarepulse(sig, duty: float = 0.5):
+    """
+    compute squarepulse using pytorch
+    equivalent to numpy implementation from
+    https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.square.html
+    """
+    t, w = torch.tensor(sig), torch.tensor(duty)
+    w = torch.tensor(w + (t - t))
```

Comment:
Is the `t - t` part here needed? If there's a rationale please comment on it. Also using `torch.tensor` directly throws up the warning `To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).` You can use `torch.as_tensor` when making tensors for arguments but on this line you don't need anything, `w+(t-t)` produces a tensor. If you did want to copy then `clone` is what to use.

Reply:
indeed (t - t) is 100% useless :-) and thanks for the hints on tensor stuffs
By the way, is the location of the functions `check_boundaries`, `paste`, `paste_slices` and `squarepulse` correct or is there a better place to store them in ? Those are helper functions for the signal transforms, I don't think they can be used for anything else

Reply:
For now we can have these functions here, `utils.py` is getting rather large so we should consider a refactor in the future however. Your functions are specific to your transforms but it's possible you could use existing general purpose functions to do the same job (`copypaste_arrays` perhaps could work in place on your `paste`). I wouldn't worry too much about that now since again we can refactor functions in the future to reduce code.

Reply:
right thanks. Last thing, as I m looking at fourier transform, I realize that in fact we could use the existing function in `torchaudio` , is `torchaudio` considered as part of monai "core", as it is part of the pytorch ecosystem or should I consider it as an optional dependency here ?

Reply:
@ericspod back on this topic ... I think I can simply remove this STFT transform for now as again there is already an equivalent transform in torchaudio and we can easily chain it with the signal transforms like for instance :

`tfs = Compose([SignalRandAddGaussianNoise(boundaries=(0.001,0.02)),
                Spectrogram(n_fft=64,
                win_length=64,
                hop_length=32,
                center=True,
                pad_mode="reflect",
                power=1.0,
                normalized=True)
               ])`

Reply:
You can if you like but we'd be adding torchaudio as an optional dependency rather than having everything inclusive in our own library.